lib/WayfProvider.php

```diff
@@ -0,0 +1,120 @@
+use OCA\Contacts\Service\FederatedInvitesService;
+use OCP\Http\Client\IClientService;
+use OCP\IAppConfig;
+use OCP\IURLGenerator;
+use OCP\OCM\IOCMDiscoveryService;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Class WayfProvider.
+ * Provides a basic WAYF (Where Are You From) login implementation build from the list of available mesh providers.
+ *
+ */
+class WayfProvider implements IWayfProvider {
+
+	public function __construct(
+		private IAppConfig $appConfig,
+		private IClientService $httpClient,
+		private FederatedInvitesService $federatedInvitesService,
+		private LoggerInterface $logger,
+		private IOCMDiscoveryService $discovery,
+		private IURLGenerator $urlGenerator,
+	) {
+	}
+
+	/**
+	 * Returns all mesh providers.
+	 *
+	 * @return array an array containing all mesh providers
+	 */
+
+
+	public function getMeshProviders(): array {
+		$urls = preg_split('/\s+/', trim($this->appConfig->getValueString(Application::APP_ID, 'mesh_providers_service')));
+		$federations = [];
+
+		$found = [];
+		foreach ($urls as $url) {
+			if ($url === '') {
+				continue;
+			}
+			try {
+				$res = $this->httpClient->newClient()->get($url);
+				$code = $res->getStatusCode();
+				if (!($code >= 200 && $code < 400)) {
+					continue;
+				}
+				$data = json_decode($res->getBody(), true);
+				$fed = $data['federation'] ?? 'Unknown';
+				$federations[$fed] = $federations[$fed] ?? [];
+
+				foreach ($data['servers'] as $prov) {
+					$fqdn = parse_url($prov['url'], PHP_URL_HOST);
+					$our_fqdn = parse_url($this->urlGenerator->getAbsoluteUrl('/'))['host'];
+					if (($our_fqdn == $fqdn) || in_array($fqdn, $found)) {
+						continue;
+					}
+					try {
+						$disc = $this->discovery->discover($prov['url'], true);
+						$inviteAcceptDialog = $disc->getInviteAcceptDialog();
+					} catch (Exception $e) {
+						$this->logger->error('Discovery failed for ' . $prov['url'] . ': ' . $e->getMessage(), ['app' => Application::APP_ID]);
+						continue;
+					}
+					if ($inviteAcceptDialog === '') {
+						// We fall back on Nextcloud default path
+						$inviteAcceptDialog = $prov['url'] . $this::INVITE_ACCEPT_DIALOG;
+					}
+					$federations[$fed][] = [
+						'provider' => $disc->getProvider(),
+						'name' => $prov['displayName'],
+						'fqdn' => $fqdn,
+						'inviteAcceptDialog' => $inviteAcceptDialog,
+					];
+					array_push($found, $fqdn);
+				}
+				usort($federations[$fed], fn ($a, $b) => strcmp($a['name'], $b['name']));
+			} catch (Exception $e) {
+				$this->logger->error('Fetch failed for ' . $url . ': ' . $e->getMessage(), ['app' => Application::APP_ID]);
+			}
+		}
+		return $federations;
+	}
+	/**
+	 * Returns all mesh providers from cache if possible.
+	 *
+	 * @return array an array containing all mesh providers
+	 */
+
+
+	public function getMeshProvidersFromCache(): array {
+		$json = $this->appConfig->getValueString(Application::APP_ID, 'federations_cache');
+		$data = json_decode($json, true);
```

Comment:
you can use getValueArray() / setValueArray() and set the config value as `lazy`